src/Native/System.Security.Cryptography.Native/pal_ecdsa_import_export.cpp

```diff
@@ -0,0 +1,431 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#include <openssl/ec.h>
+#include "pal_ecdsa_import_export.h"
+#include "pal_utilities.h"
+
+ECCurveType MethodToCurveType(EC_METHOD* method)
+{
+	if (method == EC_GFp_mont_method())
+		return ECCurveType::PrimeShortWeierstrass;// PrimeMontgomery;
+
+	if (method == EC_GF2m_simple_method())
+		return ECCurveType::Characteristic2;
+
+	// fallback including nist curves (EC_GFp_nist_method and others)
+	return ECCurveType::PrimeShortWeierstrass;
+}
+
+EC_METHOD* CurveTypeToMethod(ECCurveType curveType)
+{
+	if (curveType == ECCurveType::PrimeShortWeierstrass)
+		return const_cast<EC_METHOD*>(EC_GFp_simple_method());
+
+	if (curveType == ECCurveType::PrimeMontgomery)
+		return const_cast<EC_METHOD*>(EC_GFp_mont_method());
+
+	if (curveType == ECCurveType::Characteristic2)
+		return const_cast<EC_METHOD*>(EC_GF2m_simple_method());
+
+	return nullptr; //Edwards and others
+}
+
+extern "C" ECCurveType CryptoNative_EcKeyGetCurveType(
+	const EC_KEY* key)
+{
+	const EC_GROUP* group = EC_KEY_get0_group(key);
+	if (!group) return ECCurveType::Unspecified;
+
+	const EC_METHOD* method = EC_GROUP_method_of(group);
+	if (method == nullptr) return ECCurveType::Unspecified;
+
+	return MethodToCurveType(const_cast<EC_METHOD*>(method));
```

Comment:
In this case, that method is used elsewhere with a 'goto error' pattern which requires variables to be defined prior to the first goto, so they couldn't be const in the other case. If I change it here, I would need to undo the const elsewhere. I make look at cleaning up some of the gotos to avoid this.